sdk/nodejs/tests/provider/experimental/testdata/map-types/index.ts

```diff
@@ -0,0 +1,22 @@
+// Copyright 2025-2025, Pulumi Corporation.  All rights reserved.
+
+import * as pulumi from "@pulumi/pulumi";
+
+export interface MyComponentArgs {
+    aRecordOfStrings: Record<string, string>;
+    aRecordOfNumbers?: Record<string, number>;
+    aRecordOfBooleans: Record<string, boolean>;
+    aMapOfStrings: { [key: string]: string };
+    aMapOfNumbers: { [key: string]: number };
+    aMapOfBooleans?: { [key: string]: boolean };
+}
+
+export class MyComponent extends pulumi.ComponentResource {
+    outMapOfStrings: pulumi.Output<Map<string, string>>;
+    outMapOfNumbers: pulumi.Output<Map<string, number>>;
+    outMapOfBooleans: pulumi.Output<Map<string, boolean>>;
```

Comment:
Similarly, we probably want test cases like

```
pulumi.Output<{[key: string]: string}>;
```

and also optional

```
pulumi.Output<{[key: string]: string} | undefined>;
```

Reply:
It’s not the same though. The optionally isn’t known at preview time, so it has to live inside the Output, also to enable apply.

Reply:
Makes sense.  Union types are currently under the cut line for v1, should we pull them up?  Either way they seem kinda unrelated to maps, so I'd want to implement them in a separate PR either way.

Reply:
I don't think it's a union type - it's an output property that isn't marked as required?

Reply:
Yeah that makes sense.  I've been trying to solve this, but am currently struggling with the compiler API, that won't give me the union type when doing `<type> | undefined`.  I think we can merge this PR independently of that though, since this isn't really a concern for maps, but also any other type, e.g. `pulumi.Output<string | undefined>` or any other type.

Reply:
Finally figured out how to do this.  Draft PR in https://github.com/pulumi/pulumi/pull/18654 (it depends on this PR, so will mark it as ready for review as soon as this is merged)